File: Chapter03/3.4 Refactor OrderProcessor/GoodWay/OrderServiceComponents.cs
namespace Refactor.GoodWay
{
    // --- Step 1: The Individual Service Classes ---

    // Handles only validation logic
    public class OrderValidator
    {
        public void Validate(Order order)
        {
            Console.WriteLine("(GoodWay) Validating order...");
            if (order.Items.Count == 0 || order.Total <= 0)
            {
                throw new InvalidOperationException("Order is invalid.");
            }
        }
    }

    // Handles only payment processing
    public class PaymentService
    {
        public bool ProcessPayment(Order order)
        {
            Console.WriteLine($"(GoodWay) Processing payment for ${order.Total}...");
            // Real payment gateway logic would go here
            return true;
        }
    }

    // Handles only inventory updates
    public class InventoryManager
    {
        public void UpdateInventory(Order order)
        {
            Console.WriteLine("(GoodWay) Updating inventory...");
            // Real database logic to update stock would go here
        }
    }

    // Handles only sending notifications
    public class NotificationService
    {
        public void SendConfirmationEmail(Order order)
        {
            Console.WriteLine($"(GoodWay) Sending confirmation email to {order.CustomerEmail}...");
            // Real email sending logic would go here
        }
    }
}

File: Chapter 03/Python/3.4 OrderProcessor Refactor/After/order_service_components.py
# --- Step 1: The Individual Service Classes ---

# Handles only validation logic
class OrderValidator:
    def validate(self, order):
        print("  [Validate] Validating order...")
        if not order.items or order.total <= 0:
            raise ValueError("Order is invalid.")

# Handles only payment processing
class PaymentService:
    def process_payment(self, order):
        print(f"  [Payment] Processing payment for ${order.total:.2f}...")
        # Real payment gateway logic would go here
        return True

# Handles only inventory updates
class InventoryManager:
    def update_inventory(self, order):
        print("  [Inventory] Updating inventory...")
        # Real database logic to update stock would go here

# Handles only sending notifications
class NotificationService:
    def send_confirmation_email(self, order):
        print(f"  [Notify] Sending confirmation email to {order.customer_email}...")
        # Real email sending logic would go here